### PR TITLE
Controller class for create playlist added & modification of use case done

### PR DESCRIPTION
**Pull request by NicholasMacasaet**
Modified use case to save the newly created playlist to the User's playlist list & created controller class for this use case. Although, unsure if for the Controller, if this is the layer that needs to send a string notification to the presenter so I would like some feedback on this

### CODE REVIEW

**Review by rafaelgacesa — DISMISSED** [review]
The casting is a bit awkward and something to maybe discuss at an office hour but everything else seems fine at the moment.

File: src/main/java/Controller/newPlaylistController.java
Comment (sarxhlwalker) [inline src/main/java/Controller/newPlaylistController.java:36]: If you make the owner parametera  RegularUser (I believe artistUsers _cannot_ own playlists), then that should remove the need for casting in createPlaylist.java, no?

**Review by sarxhlwalker — COMMENTED** [review]
See comment.

File: src/main/java/Controller/newPlaylistController.java
Comment (NicholasMacasaet) [reply src/main/java/Controller/newPlaylistController.java:36]: That is true but im also unsure whether or not the parameters for this should be the id, owner and name or the input data

Comment (NicholasMacasaet) [reply src/main/java/Controller/newPlaylistController.java:36]: I looked at your implementation for the controller and I saw that

**Comment by NicholasMacasaet** [comment]
The view model is most definitely incorrect, but if possible I'd like feedback about whether my modification of the use case to pass the playlist notification message as an output data object is correct

File: src/main/java/ViewModel/playlistViewModel.java
Comment (sarxhlwalker) [inline src/main/java/ViewModel/playlistViewModel.java:16]: This requires an input data variable, based on the way you wrote newPlaylistController.java. In actuality though, I believe the controller is supposed to create the input data, not get it from the ViewModel. I believe this is why your autograde fails.

**Review by sarxhlwalker — CHANGES_REQUESTED** [review]

File: src/main/java/ViewModel/playlistViewModel.java
Comment (sarxhlwalker) [reply src/main/java/ViewModel/playlistViewModel.java:16]: I don't think you need a newPlaylistInputBoundary either, because IBs are supposed to wrap controllers -- why would you pass a controller to the construction of a new controller?

File: src/main/java/ViewModel/playlistViewModel.java
Comment (NicholasMacasaet) [inline src/main/java/ViewModel/playlistViewModel.java:24]: Not entirely sure if this is legal but I implemented the presenter/view stuff according to how Sarah described

File: src/main/java/Controller/newPlaylistController.java
Comment (sarxhlwalker) [inline src/main/java/Controller/newPlaylistController.java:11]: Again, I was under the impression that the controller should be constructing the use case interactor instead of being given it from the viewModel. if this is incorrect, please let me know.

File: src/main/java/InputData/newPlaylistInputData.java
Comment (sarxhlwalker) [inline src/main/java/InputData/newPlaylistInputData.java:9]: If you want, you could add a comment explaining why the owner is a RegularOwner (instead of User) is a valid assumption to make.

File: src/main/java/Presenter/playlistPresenter.java
Comment (sarxhlwalker) [inline src/main/java/Presenter/playlistPresenter.java:10]: I assume this is not meant to be default and instead private?

File: src/main/java/Presenter/playlistPresenter.java
Comment (sarxhlwalker) [inline src/main/java/Presenter/playlistPresenter.java:28]: Why are you passing newPlaylistOutputData into the function and then not using it? Presumably viewModel.callNewPlaylistUseCase() should take newPlaylistOutputData as an input, then.

File: src/main/java/ViewModel/playlistViewModel.java
Comment (sarxhlwalker) [inline src/main/java/ViewModel/playlistViewModel.java:33]: I'm a bit confused as to what you're trying to do in this function. Could you add comments? 

Also, see my comment on playlistPresenter -- if you pass the string message in from the Presenter as a parameter, you can just return it immediately.

File: src/main/java/UseCase/createPlaylist.java
Comment (sarxhlwalker) [inline src/main/java/UseCase/createPlaylist.java:13]: You technically don't need an outputData attribute, as you're creating it and then immediately passing it to the OutputBoundary/Presenter.

File: src/main/java/UseCase/createPlaylist.java
Comment (sarxhlwalker) [inline src/main/java/UseCase/createPlaylist.java:43]: If I'm understanding your code + clean architecture correctly -- and if I'm not, let me know! -- I don't think you need to directly return the string. You should store the OutputBoundary/Presenter as an attribute and then call its function with the outputData (i.e. the string) as a parameter.

File: src/main/java/ViewModel/playlistViewModel.java
Comment (sarxhlwalker) [inline src/main/java/ViewModel/playlistViewModel.java:19]: I don't think you should construct the OutputData in the presenter. I could be wrong, however.

Comment (sarxhlwalker) [reply src/main/java/ViewModel/playlistViewModel.java:19]: If that's the case, Presenter does not need to take OutputData as a parameter.

File: src/main/java/ViewModel/playlistViewModel.java
Comment (sarxhlwalker) [inline src/main/java/ViewModel/playlistViewModel.java:22]: Similarly, I don't believe you should construct the use case interactor in the presenter. This should occur within the controller.

**Review by sarxhlwalker — COMMENTED** [review]
I left a lot of comments, but please take what I wrote with a grain of salt. I may have misunderstood things based on our discussion today. I'm not sure whether the creation of InputData should occur in the ViewModel or the Controller (they're both in the same green layer), because my use case is so different. I'm mildly more confident in my other comments, however.

File: src/main/java/InputData/newPlaylistInputData.java
Comment (NicholasMacasaet) [reply src/main/java/InputData/newPlaylistInputData.java:9]: Will do

File: src/main/java/UseCase/createPlaylist.java
Comment (NicholasMacasaet) [reply src/main/java/UseCase/createPlaylist.java:13]: Thats true, I think I made my code significantly more complicated by trying to use output data

File: src/main/java/ViewModel/playlistViewModel.java
Comment (NicholasMacasaet) [reply src/main/java/ViewModel/playlistViewModel.java:22]: Ok, that's what I think was missing

File: src/main/java/UseCase/createPlaylist.java
Comment (sarxhlwalker) [reply src/main/java/UseCase/createPlaylist.java:43]: I understand that what i wrote might be a bit confusing! To try to make things clearer, i think what i'm envisioning is your createPlaylist stores the presenter/outputboundary as an instance attribute. then in the method, construct the outputdata, then call the presenter's function. you don't need to store the outputdata as an instance attribute itself, but you could if you want to.

File: src/main/java/Controller/newPlaylistController.java
Comment (sarxhlwalker) [inline src/main/java/Controller/newPlaylistController.java:29]: IIRC the diagram suggests the use case should take the presenter as a parameter (as an OutputBoundary) and call the presenter's function from within the use case interactor, but this technically does not break clean architecture. 

I'd prefer we keep the places where the presenter is called as consistent as possible, however.

File: src/main/java/InputData/newPlaylistInputData.java
Comment (sarxhlwalker) [inline src/main/java/InputData/newPlaylistInputData.java:20]: Similarly, you could add a comment here about why it's a RegularUser.

File: src/main/java/OutputData/editPlaylistOutputData.java
Comment (sarxhlwalker) [inline src/main/java/OutputData/editPlaylistOutputData.java:4]: is there a reason that these are default? if you have public getters, you could have these as private.

File: src/main/java/OutputData/newPlaylistOutputData.java
Comment (sarxhlwalker) [inline src/main/java/OutputData/newPlaylistOutputData.java:3]: same as above

File: src/main/java/UseCase/createPlaylist.java
Comment (sarxhlwalker) [inline src/main/java/UseCase/createPlaylist.java:11]: if you could remove commented out imports / functions, that would be great as it would make reading your code a lot easier.

File: src/main/java/UseCase/createPlaylist.java
Comment (sarxhlwalker) [inline src/main/java/UseCase/createPlaylist.java:53]: why is the outputBoundary / presenter returning information to the use case? it should be fine to call the presenter's function and not store the resulting message in the use case interactor.

**Review by sarxhlwalker — CHANGES_REQUESTED** [review]
Looks good! Please see the outstanding comments.

File: src/main/java/OutputData/editPlaylistOutputData.java
Comment (NicholasMacasaet) [reply src/main/java/OutputData/editPlaylistOutputData.java:4]: The edit playlist stuff I created just to have something implemented, I updated those variable to be private now.

File: src/main/java/Controller/newPlaylistController.java
Comment (NicholasMacasaet) [reply src/main/java/Controller/newPlaylistController.java:29]: That method I changed in the most recent commit I'm going to push now, so it should be better.

**Comment by NicholasMacasaet** [comment]
Going to go to Paul Gries' office hours tomorrow to get feedback on my use case, if everything is good I'll implement editPlaylist almost exactly how I implemented this use case.

**Comment by sarxhlwalker** [comment]
> Going to go to Paul Gries' office hours tomorrow to get feedback on my use case, if everything is good I'll implement editPlaylist almost exactly how I implemented this use case.

Thanks for the quick updates. I'll re-review this ASAP, but for ease of reviewing, could you either implement editPlaylist in a different branch / PR, or wait to add those commits to this branch until this has been merged? I expect / hope we'll be able to merge this into main soon, because it's been open for a while now.

**Comment by NicholasMacasaet** [comment]
> > Going to go to Paul Gries' office hours tomorrow to get feedback on my use case, if everything is good I'll implement editPlaylist almost exactly how I implemented this use case.
> 
> Thanks for the quick updates. I'll re-review this ASAP, but for ease of reviewing, could you either implement editPlaylist in a different branch / PR, or wait to add those commits to this branch until this PR has been merged? I expect / hope we'll be able to merge this into main soon, because it's been open for a while now.

If you do do some more reviewing could you wait until tomorrow if possible ? (check your discord for said reason). And of course I'll wait until this branch is merged. In regards to the length of time this has been open that is my fault due to my previously lacking understanding with how the view model and presenter stuff fit in.

**Comment by NicholasMacasaet** [comment]
After getting feedback from Gries he said my code was pretty good and all I needed to do was just move some stuff around. I'm ready to code edit playlist when this goes through

File: src/main/java/Controller/newPlaylistController.java
Comment (chrstinalin) [inline src/main/java/Controller/newPlaylistController.java:19]: This isn't technically wrong, but for the sake of consistency I believe we were making the controller construct the interactor.

Comment (chrstinalin) [reply src/main/java/Controller/newPlaylistController.java:19]: You'd probably also want to take in the Presenter as an OutputBoundary (as you had commented out originally) to give to the Interactor

File: src/main/java/Presenter/playlistPresenter.java
Comment (chrstinalin) [inline src/main/java/Presenter/playlistPresenter.java:10]: RE: Sarah

File: src/main/java/UseCase/CreatePlaylistInteractor.java
Comment (chrstinalin) [inline src/main/java/UseCase/CreatePlaylistInteractor.java:15]: Not wrong at all but if you wanted to make this clearer to yourself you could rename outputBoundary to Presenter (since thats what would be passed).

File: src/main/java/UseCase/CreatePlaylistInteractor.java
Comment (chrstinalin) [inline src/main/java/UseCase/CreatePlaylistInteractor.java:12]: private final?

File: src/main/java/UseCase/CreatePlaylistInteractor.java
Comment (chrstinalin) [inline src/main/java/UseCase/CreatePlaylistInteractor.java:25]: (Mostly for consistency) Add a `private final playlistAccessInterface library;` up top and move line 25 to the initializer

File: src/main/java/UseCase/editPlaylist.java
Comment (chrstinalin) [inline src/main/java/UseCase/editPlaylist.java:12]: Initializer?

File: src/main/java/UseCase/editPlaylist.java
Comment (chrstinalin) [inline src/main/java/UseCase/editPlaylist.java:27]: Instead of returning a String directly, you'd have passed in a Presenter in the initializer whose method you'd call

**Review by chrstinalin — CHANGES_REQUESTED** [review]
See comments

File: src/main/java/UseCase/editPlaylist.java
Comment (NicholasMacasaet) [reply src/main/java/UseCase/editPlaylist.java:12]: Ignore this for now, I was going to correctly code it later

File: src/main/java/UseCase/editPlaylist.java
Comment (NicholasMacasaet) [reply src/main/java/UseCase/editPlaylist.java:27]: This is true, I havent coded that part as I was more focused on correctly coding the create playlist use case, then updating edit playlist to do the same

File: src/main/java/Controller/newPlaylistController.java
Comment (NicholasMacasaet) [reply src/main/java/Controller/newPlaylistController.java:19]: Ah, I might have missed that then when re-working my code

Comment (NicholasMacasaet) [reply src/main/java/Controller/newPlaylistController.java:19]: That would require creating a new playlist interactor object in the constructor and setting this.inputboundary to said object correct ?

Comment (NicholasMacasaet) [reply src/main/java/Controller/newPlaylistController.java:19]: Also Gries' code does something similar but instead passes an input boundary object in the constructor instead of a createplaylist interactor

File: src/main/java/InputBoundary/newPlaylistInputBoundary.java
Comment (rafaelgacesa) [inline src/main/java/InputBoundary/newPlaylistInputBoundary.java:7]: This should not be a default method - this means that things that implement this interface can use this 'default' version of the method, which would be unwanted.

Instead, you should have it be public, which will also require replacing the braces with a semicolon.

File: src/main/java/Controller/newPlaylistController.java
Comment (rafaelgacesa) [inline src/main/java/Controller/newPlaylistController.java:14]: This is assuming that we will be passing an interactor into the controller, which is not the way we've been working in previous controllers (thus far we have been constructing them inside the controller). The other option is to construct the interactor in the view model, which it appears is what you are doing. While this is valid, and will work (and might even be the better choice tbh, now that I think about it), perhaps we should talk about this as a group and decide on one method or the other for consistency across our codebase.

Comment (rafaelgacesa) [reply src/main/java/Controller/newPlaylistController.java:14]: For now, no changes need to be made - it works and is a 'correct' implementation by clean architecture standards. I think I'll even lobby for changing our other controller implementations to follow this one - it's less clutter since we don't have to ever give the controller the presenter. Thanks for making me realize this lol

File: src/main/java/UseCase/CreatePlaylistInteractor.java
Comment (rafaelgacesa) [inline src/main/java/UseCase/CreatePlaylistInteractor.java:38]: This is a great implementation and correct use of the playlist database. However, I am wondering if you will also have to use the user database class to save the user, after assigning them the new playlist.

I think you will have to do this, since there is no connection between the RegularUser class and the user database to say, "Hey, I have a new playlist that needs to be added to my data and saved!". So, I think you'll have to add the instance variable `userAccessInterface userDatabase` to the top of your file, and then instantiate it similarly to the playlist library, and then call the `userDatabase.save()` method on the `RegularUser` after you add the playlist to it in line 38. This will ensure that the new playlist is correctly written to the database as belonging to the user.

File: src/main/java/ViewModel/playlistViewModel.java
Comment (rafaelgacesa) [inline src/main/java/ViewModel/playlistViewModel.java:10]: This is the method Gries likes, since he really focuses on dependency injection. However, we've been created the controllers and presenters inside of the view model, which is definitely allowed according to clean architecture. I've also personally asked Prof. Gries about this and he said that it is fine to do this too, he just prefers dependency injection. 

I think we should stick to constructing our controllers and presenters inside of the view model, just to keep our code consistent, even though it introduces some hard dependencies. If we have time at the end of development, we can take a step back and see if we can apply dependency injection anywhere to make things cleaner, but for now we should be consistent.

**Review by rafaelgacesa — CHANGES_REQUESTED** [review]
Overall looks like good progress on your use case, no major errors. Some minor comments regarding small changes and things to think about.

**Review by rafaelgacesa — APPROVED** [review]
Thanks for implementing the changes! I got a bit confused about the input boundary (since it's empty now) but I realized that's the one for the use case you haven't implemented yet. This pull looks good to go!

**Review by Mistryr222 — APPROVED** [review]
Looks Good!

**Review by chrstinalin — APPROVED** [review]
Looks good!

**Pull request merged**
